### PR TITLE
[udp6] update UDP socket definitions

### PR DESCRIPTION
**Pull request by abtink**
This commit updates the public OT `otUdp{}` (like `otUdpConnect` or `otUpdSend()`) functions to requires a pointer the OpenThread instance to be passed as their first parameter (this harmonizes the definitions with other public APIs and removes the need for `otUdpSocket` to track the instance).

The core implementation in `udp6` module is updated to define `SocketHandle` as an internal mirror of `otUdpSocket` structure. The socket related methods `Open`(), `Bind()`, `Connect()`, etc., are moved to `Udp` class itself (with a `SocketHandle` passed as a parameter). For core internal use `Socket` class is defined as a sub-class of `SocketHandle` and `InstanceLocator` providing same helper methods. The separation between `SocketHandle` and `Socket` type addresses the problem that can be caused by the treating/casting publicly provided `otUdpSockt` objects as `InstanceLocator`.

--------

This should help address issue https://github.com/openthread/openthread/issues/5162.

### CODE REVIEW

File: src/core/net/udp6.hpp [hunk 1]
```diff
@@ -348,20 +358,66 @@ class Udp : public InstanceLocator
...
     /**
-     * This method adds a UDP socket.
+     * This method opens a UDP socket.
+     *
+     * @param[in]  aSocket   A reference to the socket.
+     * @param[in]  aHandler  A pointer to a function that is called when receiving UDP messages.
+     * @param[in]  aContext  A pointer to arbitrary context information.
+     *
+     * @retval OT_ERROR_NONE     Successfully opened the socket.
+     * @retval OT_ERROR_ALREADY  The socket is already open.
```
Comment (jwhui) [inline src/core/net/udp6.hpp:368]: Replace with `OT_ERROR_FAILED`?

File: src/core/net/udp6.hpp [hunk 2]
```diff
@@ -348,20 +358,66 @@ class Udp : public InstanceLocator
...
-     * This method removes a UDP socket.
+     * This method sends a UDP message using a socket.
      *
-     * @param[in]  aSocket  A reference to the UDP socket.
+     * @param[in]  aSocket       A reference to the socket.
+     * @param[in]  aMessage      The message to send.
+     * @param[in]  aMessageInfo  The message info associated with @p aMessage.
+     *
+     * @retval OT_ERROR_NONE          Successfully sent the UDP message.
+     * @retval OT_ERROR_INVALID_ARGS  If no peer is specified in @p aMessageInfo or by connect().
```
Comment (jwhui) [inline src/core/net/udp6.hpp:416]:
```suggestion
     * @retval OT_ERROR_INVALID_ARGS  If no peer is specified in @p aMessageInfo or by Connect().
```

File: include/openthread/udp.h [hunk 3]
```diff
@@ -157,68 +157,51 @@ otMessage *otUdpNewMessage(otInstance *aInstance, const otMessageSettings *aSett
...
+ * @retval OT_ERROR_FAILED Failed to connect UDP socket.
  *
  */
-otError otUdpConnect(otUdpSocket *aSocket, otSockAddr *aSockName);
+otError otUdpConnect(otInstance *aInstance, otUdpSocket *aSocket, const otSockAddr *aSockName);
 
 /**
  * Send a UDP/IPv6 message.
  *
+ * @param[in]  aInstance  A pointer to an OpenThread instance.
```
Comment (jwhui) [inline include/openthread/udp.h:204]:
```suggestion
 * @param[in]  aInstance     A pointer to an OpenThread instance.
```

**Comment by romacdon** [comment]
@abtink and @jwhui - this commit breaks the efr32 builds.  I'll have someone fix up the efr32 builds but whats the normal policy on making an incompatible public API change like this?

**Comment by abtink** [comment]
@romacdon we track any public OT API changes (even a change in a documentation) with an API version number increase (this is away for users of OT APIs to be aware of such changes) -> https://github.com/openthread/openthread/blob/b88f4fccba5bd00c4f49bc6bc0d77ebc3d23a2ad/include/openthread/instance.h#L48-L56

We also have a CI test/check which ensures any PRs that change any public OT API source files, does indeed update the API version number as well.

**Comment by romacdon** [comment]
Thanks @abtink - I didn't know about the rules around OPENTHREAD_API_VERSION. Good to know thats in place.  Was there any thought about deprecating old APIs before actually removing them?

**Comment by abtink** [comment]
We generally do not remove OT public APIs. We may add/enhance or fix some of the APIs. 

For example, this PR addressed the issue where some of the `otUdp` APIs that they did not provide an `otInstance *` as a parameter - which is typically done/expected for all OT public APIs. This could lead to issues in some situations/configs, explained here #5162.

**Comment by romacdon** [comment]
I've worked on projects before that even when APIs are enhanced or fixed the old signature of the API is still supported for some time.  Anyway, I'm OK with this change and will have someone address the issues introduced by this change.  Thanks for your explanations, @abtink .

**Comment by abtink** [comment]
@romacdon sounds good. Thanks.

Is the efr32 example code (using this API) part of OT repo itself? If so, I highly suggest adding CI tests for it. This would help detect any change that would break the example code. I would have fixed/updated it as part of this PR itself, if I knew that the code was impacted and was part of OT repo.

**Comment by romacdon** [comment]
@abtink yes the code is part of the OT repo.  The difficulty of adding CI tests is the inclusion of the Silabs SDK.  Until that is in an easily downloadable location its won't be possible to add the CI tests.  We are working on making it easily downloadable.

**Pull request merged**